### ISSUE TITLE
Add ToTypedArray method in Godot.Collections.Array C#

### ISSUE THREAD

**Issue by ghsoares** [godotengine/godot-proposals#1177]
**Describe the project you are working on:**
A C# project

**Describe the problem or limitation you are having in your project:**
Many of C# methods implemented by the engine returns Godot.Collections.Array and there's no intuitive way to convert to typed array. I can't just use the type casting like "(Vector2[])array" to get the typed version, or convert to Godot.Collections.Array<Vector2> then to Vector2[].

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Implement a simple function that returns a typed array to Godot.Collections.Array class so we can use these instead.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
In the class could be something like this:

```
public T[] ToTypedArray<T>() {
    T[] arr = new T[this.Count];
    // Some magic to store the values in the array
    return arr;
}
```

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Yes, we can create a simple function with foreach loop to store the Array values in the typed array, but this isn't intuitive in the perspective of the developer.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Because simple and useful premade functions like that engages game development in C# in Godot.

**Comment by raulsntos** [comment]
`Godot.Collections.Array` implements `System.Collections.IEnumerable` so you could use [`Cast<T>()`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.cast) or [`OfType<T>()`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.oftype) on `IEnumerable` to convert it into `IEnumerable<T>` like this:
```csharp
IEnumerable<Node> children = GetChildren().Cast<Node>();
IEnumerable<Label> labels = GetChildren().OfType<Label>();
```
If you use `Cast<T>` make sure all elements are compatible with `T`, otherwise you will get `System.InvalidCastException`, to avoid this you could use `OfType<T>` instead so if there aren't any compatible elements with type `T` it will return an empty array and no Exceptions.
Then, if you really want it to be an array `System.Linq` allows you to call `ToArray()` on an `IEnumerable<T>`.

---

@neikeq This question was asked before in godotengine/godot#35182 so it seems it's not very clear that you can already convert Godot's collections to a typed C# collection. In 3.1 `Array` implemented `IEnumerable<object>` instead of the non-generic `IEnumerable` but it was changed in PR godotengine/godot#26662, I'm not sure why it was changed but I wonder if it should be changed back if it will make it easier or less confusing to work with.

**Comment by ghsoares** [comment]
@raulsntos indeed, maybe before that could work but I don't know why the method Cast<T> isn't a definition to godot array, even if implements IEnumerable, maybe this error is because of my dotnet version (4.7.2), but I aint sure.

**Comment by ghsoares** [comment]
@raulsntos Nervermind, I forgot to add `using System.Linq` to the header of the document but maybe, just maybe it could be nice to have this implementation directly to Godot Array

**Comment by EricEzaM** [comment]
How would godots implementation of this be different to that in C#? If you are using C#, then you might as well use the language features it provides.

**Comment by cgbeutler** [comment]
Using Linq's `ToArray` will convert to the `System.Array` type instead of Godot's array type.
I would argue that, if this method is added, it should convert to `Godot.Collections.Array<>` instead of `T[]`.
I would also add that having a way to get from `Godot.Collections.Dictionary` to `Godot.Collections.Dictionary<,>` without having to go some long way would be nice.

**Comment by neikeq** [comment]
There's `CopyTo`. Is this what this proposal is asking for?

```
Godot.Collections.Array<Vector2> godotArray = Foo();
Vector2[] array = new Vector2[godotArray.Count];
godotArray.CopyTo(array, 0);
```

**Comment by raulsntos** [comment]
@cgbeutler

> I would argue that, if this method is added, it should convert to `Godot.Collections.Array<>` instead of `T[]`.

`Godot.Collections.Array<>` already has a constructor that takes an untyped `Godot.Collections.Array`:

https://github.com/godotengine/godot/blob/aaf556685da905857326b2d6ab41360f095a7596/modules/mono/glue/GodotSharp/GodotSharp/Core/Array.cs#L396

So you can just do this:

```csharp
Godot.Collections.Array untypedArray = Foo();
Godot.Collections.Array<Vector2> typedArray = new Godot.Collections.Array<Vector2>(untypedArray);
```

> I would also add that having a way to get from `Godot.Collections.Dictionary` to `Godot.Collections.Dictionary<,>` without having to go some long way would be nice.

`Godot.Collections.Dictionary<>` also has a constructor that takes an untyped `Godot.Collections.Dictionary`:

https://github.com/godotengine/godot/blob/aaf556685da905857326b2d6ab41360f095a7596/modules/mono/glue/GodotSharp/GodotSharp/Core/Dictionary.cs#L414

So you can just do this:

```csharp
Godot.Collections.Dictionary untypedDictionary = Foo();
Godot.Collections.Dictionary<string, Vector2> typedDictionary = new Godot.Collections.Array<string, Vector2>(untypedDictionary);
```

> It's tricky, though, since Godot has Array and Dictionary as taking variants.

If the collection contains an element of a different type than the specified in the typed collection it will throw an [InvalidCastException](https://docs.microsoft.com/en-us/dotnet/api/system.invalidcastexception?view=net-5.0) when retrieving a value from the collection since it tries to cast the `Variant`.

To get rid of the non-compatible types you can use `System.Linq`'s `OfType<T>` which returns a `IEnumerable<T>` that only contains the elements that can be casted to `T` and then `System.Linq` has `ToArray()` which converts it to a `T[]` like this proposal requested, or if you want a `Godot.Collections.Array<T>` you can use the constructor that takes `IEnumerable<T>`.

---

@neikeq `CopyTo` should also satisfy what the proposal is asking for and then there's no need to use `System.Linq`.
